### ISSUE TITLE
Condition does not work, neither if not switch

### ISSUE THREAD

**Issue by xperiandri** [dotnet/roslyn#12915]
**Version Used**: Visual Studio 2015 Update 3

**Steps to Reproduce**:
I can give you sources and binaries.

**Expected Behavior**:
Condition is preserved and conditional code is not executed

**Actual Behavior**:
<img width="526" alt="true false" src="https://cloud.githubusercontent.com/assets/2365592/17383820/b77a81b6-59e0-11e6-942c-0350676cee5d.png">
<img width="610" alt="true false2" src="https://cloud.githubusercontent.com/assets/2365592/17383823/ba2f3df2-59e0-11e6-9e48-29d652c57a39.png">

**Comment by svick** [comment]
> I can give you sources and binaries.

Please do, ideally in the form of a [short, self-contained example](http://sscce.org/). Without the source it's usually nigh impossible to diagnose the problem.

---

Also, are you sure the code is actually executed and it's not just debugger showing you the wrong line?

**Comment by gafter** [comment]
Please provide a self-contained program whose execution is incorrect.

If you are unable to do that, but you still experience the problem in the debugger, please provide the program whose positions are displayed incorrectly in the debugger.

**Comment by gafter** [comment]
Here are two snippets of code that illustrate the problem precisely as described:

```cs
    public static async Task M1()
    {
        var task = Task.FromResult(true);
        bool b1 = await task;
        if (!b1)
        {
            throw null;
        }
    }
```

```cs
    public static async Task M2()
    {
        var task = Task.FromResult(true);
        bool b2 = await task;
        switch (b2)
        {
            case false:
                break;
        }
    }
```

This appears to be a problem with the interaction between the sequence points and the async rewriter at the end of an async method.

**Comment by jcouv** [comment]
Relates to https://github.com/dotnet/roslyn/issues/12564

**Comment by jcouv** [comment]
@xperiandri Just to clarify, despite what the debugger shows you, the code is actually executed correctly.
 
The problem is with displaying the highlighted regions of code that the debugger is paused on. The root cause is that the compiler is not emitting the proper sequence points (which delimit such regions for the debugger).

**Comment by jcouv** [comment]
Fixed by #16803 (Missing sequence point on implicit return in async method)
The fix unfortunately didn't make it into RTW, but will be released in the next ship window.
